runtime/src/main/java/org/corfudb/protocols/wireprotocol/LogData.java

```diff
@@ -136,36 +135,60 @@ public synchronized void acquireBuffer() {
 
     @Override
     public int getSizeEstimate() {
-        byte[] tempData = data;
-        if (tempData != null) {
-            return tempData.length;
-        }
-
         if (lastKnownSize != NOT_KNOWN) {
             return lastKnownSize;
         }
-
         return 1;
     }
 
     @Getter
-    final EnumMap<LogUnitMetadataType, Object> metadataMap;
+    EnumMap<LogUnitMetadataType, Object> metadataMap;
+    private byte[] serializedMetadata;
+
+    public void serializeMetadata() {
+        if (serializedMetadata != null) {
+            return;
+        }
+        ByteBuf serializedBuf = Unpooled.buffer();
+        //System.out.println("Metadata map: " + metadataMap);
+        ICorfuPayload.serialize(serializedBuf, metadataMap);
+        serializedMetadata = byteArrayFromBuf(serializedBuf);
+        metadataMap = null;
```

Comment:
Look into byteArrayFromBuf, it is creating another buf. As the serializedBuf is fresh, you can just do : byte[] serializedMetadata = new byte[.readableBytes()];

Reply:
By doing that, I will have to repeat the same thing done in byteArrayFromBuf - 
serializedMetadata = new byte[serializedBuf.readableBytes()];
serializedBuf.readBytes(serializedMetadata);

If creating  the bytebuf is not an expensive operation, I would like to keep this unchanged.  Let me know your thoughts.